Project.toml

```diff
@@ -8,6 +8,7 @@ julia = "1"
 
 [deps]
 Requires="ae029012-a4dd-5104-9daa-d747884805df"
+BlockDiagonals="0a1fb500-61f7-11e9-3c65-f5ef3456f9f0"
```

Comment:
As my own opinion, we'd better not make non-standard libraries hard dependencies, instead, use Requires to do the optional import.

Also, this PR is growing bigger and hard to review, you can focus on fixing the existing issues in this PR, setting up a template of writing future wrappers, and leaving the `BlockDiagonals` to the next PR.

Reply:
> A second thought to this: if `blkdiag` belongs to the Matlab Core package (it is now), then we have good reason to add this package as the hard dependency.

Yes, it is part of Matlab's language. TBH, adding dependencies does not matter. Installing Matlab takes about 1 hour or something! Our adding time is far less than 1 hour.